README.md
# C64DS V0.2.0

<img align="right" width="220" src="./logo.png" />

This is a Commodore 64 (PAL) emulator for the Nintendo DS.

Too much stuff is still missing for it to be more than a novelty.
* Graphics is poorly emulated.
* Sound is missing ring, sync & filter emulation.
* Timer emulation is not cycle accurate, missing TA to TB chaining.
* Cpu is missing a lot of the finer cycle rules.
* There is no disc drive or tape emulation yet, just fake loading of ".prg" files.

## How to use

1. Create a folder named "c64ds" in either the root of your flash card or in
 the data folder. This is where settings and save files end up.
2. Now put .prg files into a folder where you have (C64) files.
3. Depending on your flashcart you might have to DLDI patch the emulator.

When the emulator starts, you can either press L+R or tap on the screen to open
up the menu. Now you can use the cross or touchscreen to navigate the menus, A
or double tap to select an option, B or the top of the screen to go back a step.

To select between the tabs use R & L or the touchscreen.

## Menu

### File

* Load Game: Select a game to load.
* Load State: Load a previously saved state of the currently running game.
* Save State: Save a state of the currently running game.
* Load NVRAM: Load non volatile ram (EEPROM/SRAM) for the currently running game.
* Save NVRAM: Save non volatile ram (EEPROM/SRAM) for the currently running game.
* Save Settings: Save the current settings (and internal EEPROM).
* Reset Game: Reset the currently running game.

### Options

* Controller:
  * Autofire: Select if you want autofire.
  * Controller: Which port the controller is plugged into, be prepared to switch often
  * Swap A-B: Swap which NDS button is mapped to the C64 fire button.
* Display:
  * Display: Here you can select if you want scaled or unscaled screenmode.
   Unscaled mode: L & R buttons scroll the screen up and down.
  * Scaling: Here you can select if you want flicker or barebones lineskip.
  * Gamma: Lets you change the gamma ("brightness").
  * Contrast: Lets you change the contrast.
  * Palettes: Here you can select the palette.
* Machine Settings:
  * Machine: Select the emulated machine.
  * Select Kernal: Load a specific Kernal ROM.
  * Select Chargen: Load a specific Chargen ROM.
  * Select Basic: Load a specific Basic ROM.
  * Cpu speed hacks: Allow speed hacks.
* Settings:
  * Speed: Switch between speed modes.
    * Normal: Game runs at it's normal speed.
    * 200%: Game runs at double speed.
    * Max: Games can run up to 4 times normal speed.
    * 50%: Game runs at half speed.
  * Allow Refresh Change: Allow the Wonderswan to change NDS refresh rate.
  * Autoload State: Toggle Savestate autoloading. Automagically load the
   savestate associated with the selected game.
  * Autoload NVRAM: Toggle EEPROM/SRAM autoloading. Automagically load the
   EEPROM/SRAM associated with the selected game.
  * Autosave Settings: This will save settings when leaving menu if any
   changes are made.
  * Autopause Game: Toggle if the game should pause when opening the menu.
  * Powersave 2nd Screen: If graphics/light should be turned off for the GUI
   screen when menu is not active.
  * Emulator on Bottom: Select if top or bottom screen should be used for
   emulator, when menu is active emulator screen is allways on top.
  * Autosleep: Doesn't work.
* Debug:
  * Debug Output: Show FPS and logged text.
  * Disable Foreground: Turn on/off foreground rendering.
  * Disable Background: Turn on/off background rendering.
  * Disable Sprites: Turn on/off sprite rendering.
  * Step Frame: Emulate one frame.

### About

Some info about the emulator and game...

## Controls

There is no convention on C64 which port is player 1 and which is player 2, so
you will have to switch port a lot. Fire is right now mapped to the B button.

## Games

* Bubble Bobble actually seem to work.
* Great Giana Sisters is missing sprite collision.

## Credits

```text
Huge thanks to Loopy for the incredible PocketNES, without it this emu would probably never have been made.
Thanks to:
Dwedit for help and inspiration with a lot of things. https://www.dwedit.org
```

Fredrik Ahlström

Twitter @TheRealFluBBa

http://www.github.com/FluBBaOfWard


History.txt
C64DS revision history
-=-=-=-=-=-=-=-=-=-=-=-

V0.2.0 - 2023-06-27 (FluBBa)
	Reworked for devkitARM.

V0.1.1 - 2023-06-04 (FluBBa)
	Started to rework for devkitARM.

V0.1.0 - 2006-xx-xx (FluBBa)
	Started coding on NDS version.


source/Gui.c
#include <nds.h>

#include "Gui.h"
#include "Shared/EmuMenu.h"
#include "Shared/EmuSettings.h"
#include "Main.h"
#include "C64Keyboard.h"
#include "FileHandling.h"
//#include "Cart.h"
#include "Gfx.h"
#include "io.h"
#include "cpu.h"
#include "ARM6502/Version.h"
#include "ARM6526/Version.h"
#include "ARM6569/Version.h"
#include "ARM6581/Version.h"

#define EMUVERSION "V0.2.0 2023-06-27"

#define ENABLE_LIVE_UI		(1<<12)
#define ALLOW_SPEED_HACKS	(1<<17)
#define ALLOW_REFRESH_CHG	(1<<19)

void machineReset(void);
void hacksInit(void);

static void nullUIC64(int key);
static void setupC64Background(void);

static void paletteChange(void);
static void machineSet(void);
static void speedHackSet(void);
static void refreshChgSet(void);
static void borderSet(void);

static void uiMachine(void);
static void uiDebug(void);
static void updateGameInfo(void);

const fptr fnMain[] = {nullUI, subUI, subUI, subUI, subUI, subUI, subUI, subUI, subUI, subUI};

const fptr fnList0[] = {uiDummy};
const fptr fnList1[] = {selectGame, loadState, saveState, loadNVRAM, saveNVRAM, saveSettings, resetGame, ui9};
const fptr fnList2[] = {ui4, ui5, ui6, ui7, ui8};
const fptr fnList3[] = {uiDummy};
const fptr fnList4[] = {autoBSet, autoASet, controllerSet, swapABSet};
const fptr fnList5[] = {scalingSet, flickSet, gammaSet, contrastSet, paletteChange, borderSet};
const fptr fnList6[] = {machineSet, selectBnWBios, selectColorBios, selectCrystalBios /*languageSet*/};
const fptr fnList7[] = {speedSet, refreshChgSet, autoStateSet, autoNVRAMSet, autoSettingsSet, autoPauseGameSet, powerSaveSet, screenSwapSet, sleepSet};
const fptr fnList8[] = {debugTextSet, bgrLayerSet, sprLayerSet, stepFrame};
const fptr fnList9[] = {exitEmulator, backOutOfMenu};
const fptr fnList10[] = {uiDummy};
const fptr *const fnListX[] = {fnList0, fnList1, fnList2, fnList3, fnList4, fnList5, fnList6, fnList7, fnList8, fnList9, fnList10};
u8 menuXItems[] = {ARRSIZE(fnList0), ARRSIZE(fnList1), ARRSIZE(fnList2), ARRSIZE(fnList3), ARRSIZE(fnList4), ARRSIZE(fnList5), ARRSIZE(fnList6), ARRSIZE(fnList7), ARRSIZE(fnList8), ARRSIZE(fnList9), ARRSIZE(fnList10)};
const fptr drawUIX[] = {uiNullNormal, uiFile, uiOptions, uiAbout, uiController, uiDisplay, uiMachine, uiSettings, uiDebug, uiDummy, uiDummy};

u8 gGammaValue = 0;
u8 gContrastValue = 3;
u8 gBorderEnable = 1;
char gameInfoString[32];

const char *const autoTxt[]  = {"Off", "On", "With R"};
const char *const speedTxt[] = {"Normal", "200%", "Max", "50%"};
const char *const brighTxt[] = {"I", "II", "III", "IIII", "IIIII"};
const char *const sleepTxt[] = {"5min", "10min", "30min", "Off"};
const char *const ctrlTxt[]  = {"Port 1", "Port 2"};
const char *const dispTxt[]  = {"Unscaled", "Scaled"};
const char *const flickTxt[] = {"No Flicker", "Flicker"};

const char *const machTxt[]  = {"Auto", "C64", "C64GS", "C64EX", "C64NTSC"};
const char *const bordTxt[]  = {"Black", "Border Color", "None"};
const char *const palTxt[]   = {"Classic", "Black & White", "Red", "Green", "Blue", "Green-Blue", "Blue-Green", "Puyo Puyo Tsu"};
const char *const langTxt[]  = {"Japanese", "English"};


void setupGUI() {
	emuSettings = AUTOPAUSE_EMULATION | AUTOLOAD_NVRAM | AUTOSLEEP_OFF | ENABLE_LIVE_UI;
	keysSetRepeat(25, 4);	// Delay, repeat.
	menuXItems[1] = ARRSIZE(fnList1) - (enableExit?0:1);
	openMenu();
}

/// This is called when going from emu to ui.
void enterGUI() {
	if (emuSettings & AUTOSAVE_SETTINGS) {
		saveSettings();
		settingsChanged = false;
	}
}

/// This is called going from ui to emu.
void exitGUI() {
}

void quickSelectGame(void) {
	openMenu();
	selectGame();
	closeMenu();
}

void uiNullNormal() {
//	if (gMachine == HW_C64) {
		setupC64Background();
//	} else {
//		uiNullDefault();
//		return;
//	}
	drawItem("Menu",27,1,0);
}

void uiFile() {
	setupMenu();
	drawMenuItem("Load Game");
	drawMenuItem("Load State");
	drawMenuItem("Save State");
	drawMenuItem("Load NVRAM");
	drawMenuItem("Save NVRAM");
	drawMenuItem("Save Settings");
	drawMenuItem("Reset Console");
	if (enableExit) {
		drawMenuItem("Quit Emulator");
	}
}

void uiOptions() {
	setupMenu();
	drawMenuItem("Controller");
	drawMenuItem("Display");
	drawMenuItem("Machine");
	drawMenuItem("Settings");
	drawMenuItem("Debug");
}

void uiAbout() {
	cls(1);
	updateGameInfo();
	drawTabs();
	drawMenuText("B:        WS B button", 4, 0);
	drawMenuText("A:        WS A button", 5, 0);
	drawMenuText("Select:   Sound button", 6, 0);
	drawMenuText("Start:    Start button", 7, 0);

	drawMenuText(gameInfoString, 9, 0);

	drawMenuText("C64DS        " EMUVERSION, 18, 0);
	drawMenuText("ARM6502      " ARM6502VERSION, 19, 0);
	drawMenuText("ARM6526      " ARM6526VERSION, 20, 0);
	drawMenuText("ARM6569      " ARM6569VERSION, 21, 0);
	drawMenuText("ARM6581      " ARM6581VERSION, 22, 0);
}

void uiController() {
	setupSubMenu("Controller Settings");
	drawSubItem("B Autofire:", autoTxt[autoB]);
	drawSubItem("A Autofire:", autoTxt[autoA]);
	drawSubItem("Controller:", ctrlTxt[(joyCfg>>29)&1]);
	drawSubItem("Swap A-B:  ", autoTxt[(joyCfg>>10)&1]);
}

void uiDisplay() {
	setupSubMenu("Display Settings");
	drawSubItem("Display:", dispTxt[gScaling&SCALED]);
	drawSubItem("Scaling:", flickTxt[gFlicker]);
	drawSubItem("Gamma:", brighTxt[gGammaValue]);
	drawSubItem("Contrast:", brighTxt[gContrastValue]);
	drawSubItem("Border:", autoTxt[gBorderEnable]);
}

static void uiMachine() {
	setupSubMenu("Machine Settings");
	drawSubItem("Machine:", machTxt[0]);
	drawSubItem("Select Kernal ->", NULL);
	drawSubItem("Select Basic ->", NULL);
	drawSubItem("Select Chargen ->", NULL);
//	drawSubItem("Language: ", langTxt[gLang]);
}

void uiSettings() {
	setupSubMenu("Settings");
	drawSubItem("Speed:", speedTxt[(emuSettings>>6)&3]);
	drawSubItem("Allow Refresh Change:", autoTxt[(emuSettings&ALLOW_REFRESH_CHG)>>19]);
	drawSubItem("Autoload State:", autoTxt[(emuSettings>>2)&1]);
	drawSubItem("Autoload NVRAM:", autoTxt[(emuSettings>>10)&1]);
	drawSubItem("Autosave Settings:", autoTxt[(emuSettings>>9)&1]);
	drawSubItem("Autopause Game:", autoTxt[emuSettings&1]);
	drawSubItem("Powersave 2nd Screen:",autoTxt[(emuSettings>>1)&1]);
	drawSubItem("Emulator on Bottom:", autoTxt[(emuSettings>>8)&1]);
	drawSubItem("Autosleep:", sleepTxt[(emuSettings>>4)&3]);
}

void uiDebug() {
	setupSubMenu("Debug");
	drawSubItem("Debug Output:", autoTxt[gDebugSet&1]);
	drawSubItem("Disable Background:", autoTxt[gGfxMask&1]);
	drawSubItem("Disable Sprites:", autoTxt[(gGfxMask>>4)&1]);
	drawSubItem("Step Frame", NULL);
}


void nullUINormal(int key) {
	if (!(emuSettings & ENABLE_LIVE_UI)) {
		nullUIDebug(key);		// Just check touch, open menu.
		return;
	}
	nullUIC64(key);
//	if (key & KEY_TOUCH) {
//		openMenu();
//	}
}

void nullUIDebug(int key) {
	if (key & KEY_TOUCH) {
		openMenu();
	}
}

void updateGameInfo() {
	char catalog[8];
//	char2HexStr(catalog, gGameID);
	strlMerge(gameInfoString, "Game #: 0x", catalog, sizeof(gameInfoString));
}
//---------------------------------------------------------------------------------
void debugIO(u16 port, u8 val, const char *message) {
	char debugString[32];

	debugString[0] = 0;
	strlcat(debugString, message, sizeof(debugString));
	short2HexStr(&debugString[strlen(debugString)], port);
	strlcat(debugString, " val:", sizeof(debugString));
	char2HexStr(&debugString[strlen(debugString)], val);
	debugOutput(debugString);
}
//---------------------------------------------------------------------------------
void debugIOUnmappedR(u16 port, u8 val) {
	debugIO(port, val, "Unmapped R port:");
}
void debugIOUnmappedW(u16 port, u8 val) {
	debugIO(port, val, "Unmapped W port:");
}
void debugUndefinedInstruction() {
	debugOutput("Undocumented Instruction.");
}
void debugCrashInstruction() {
	debugOutput("CPU Crash! (JAM/KIL)");
}

void nullUIC64(int key) {
	int xpos, ypos;

	if (EMUinput & KEY_TOUCH) {
		touchPosition myTouch;
		touchRead(&myTouch);
		xpos = (myTouch.px>>2);
		ypos = (myTouch.py>>4);
		if ( (xpos > 51) && (ypos < 2) ) {
			openMenu();
		}
		else {
			SetC64Key((myTouch.px>>3), (myTouch.py>>3), 1);
		}
//		else if ((ypos > 4 && ypos < 6) && (xpos > 21 && xpos < 43)) {	// Cartridge port
//			cartridgePortTouched(keyHit);
//		}
//		else if ((ypos == 10) && (xpos > 46 && xpos < 52)) {	// Hold button
//			EMUinput |= KEY_START;
//		}
	}
	else {
		SetC64Key(0, 0, 0);
	}
}

//---------------------------------------------------------------------------------
void setupC64Background(void) {
	setupCompressedBackground(C64KeyboardTiles, C64KeyboardMap, 9);
	memcpy(BG_PALETTE_SUB+0x80, C64KeyboardPal, C64KeyboardPalLen);
}

void resetGame() {
	machineReset();
//	loadCart();
}

//---------------------------------------------------------------------------------
/// Switch between Player 1 & Player 2 controls
void controllerSet() {				// See io.s: refreshEMUjoypads
	joyCfg ^= 0x20000000;
}

/// Swap A & B buttons
void swapABSet() {
	joyCfg ^= 0x400;
}

/// Turn on/off scaling
void scalingSet(){
	gScaling ^= SCALED;
//	refreshGfx();
}

/// Change gamma (brightness)
void gammaSet() {
	gGammaValue++;
	if (gGammaValue > 4) gGammaValue = 0;
//	paletteInit(gGammaValue, gContrastValue);
//	paletteTxAll();					// Make new palette visible
//	setupEmuBorderPalette();
	setupMenuPalette();
	settingsChanged = true;
}

/// Change contrast
void contrastSet() {
	gContrastValue++;
	if (gContrastValue > 4) gContrastValue = 0;
//	paletteInit(gGammaValue, gContrastValue);
//	paletteTxAll();					// Make new palette visible
//	setupEmuBorderPalette();
	settingsChanged = true;
}

/// Turn on/off rendering of background
void bgrLayerSet() {
	gGfxMask ^= 0x01;
}
/// Turn on/off rendering of sprites
void sprLayerSet() {
	gGfxMask ^= 0x10;
}

void paletteChange() {
/*	gPaletteBank++;
	if (gPaletteBank > 7) {
		gPaletteBank = 0;
	}
	monoPalInit(gGammaValue, gContrastValue);
	paletteTxAll();
	setupEmuBorderPalette();*/
	settingsChanged = true;
}

void borderSet() {
	gBorderEnable ^= 0x01;
//	setupEmuBorderPalette();
}

void machineSet() {
//	gMachine++;
//	if (gMachine >= HW_SELECT_END) {
//		gMachine = 0;
//	}
}

void speedHackSet() {
	emuSettings ^= ALLOW_SPEED_HACKS;
//	hacksInit();
}

void refreshChgSet() {
	emuSettings ^= ALLOW_REFRESH_CHG;
//	updateLCDRefresh();
}
